### ISSUE TITLE
Shared libraries not built if no fortran compiler is available

### ISSUE THREAD

**Issue by albestro** [pmodels/mpich#5674]
I'm building `mpich~fortran`, i.e. without fortran support, using spack on macOS with `apple-clang`.

I decided to report here the issue because I think it is not related in any way to spack.  At first I thought the problem was related to disabling fortran support, but investigating a bit more I feel the problem is in the configure process of mpich itself.

I configured spack to get two different toolchains, both with the same system compiler `apple-clang@13.0.0`, but one of the two, for which I used suffix `z` to distinguish it, has also the `fortran` compilers (using `gfortran`).

```
➜  ~ spack compiler info apple-clang
apple-clang@13.0.0:
	paths:
		cc = /usr/bin/clang
		cxx = /usr/bin/clang++
		f77 = None
		fc = None
	modules  = []
	operating system  = bigsur
apple-clang@13.0.0z:
	paths:
		cc = /usr/bin/clang
		cxx = /usr/bin/clang++
		f77 = /usr/local/bin/gfortran-11
		fc = /usr/local/bin/gfortran-11
	modules  = []
	operating system  = bigsur
```

Then I built `mpich~fortran` with both of them.

```bash
➜  ~ spack find -lv mpich
==> 2 installed packages
-- darwin-bigsur-skylake / apple-clang@13.0.0 -------------------
4vs4lt3 mpich@3.4.2~argobots~fortran+hwloc+hydra+libxml2~pci+romio~slurm+two_level_namespace~verbs+wrapperrpath device=ch3 netmod=tcp pmi=pmi

-- darwin-bigsur-skylake / apple-clang@13.0.0z ------------------
yw3oxjr mpich@3.4.2~argobots~fortran+hwloc+hydra+libxml2~pci+romio~slurm+two_level_namespace~verbs+wrapperrpath device=ch3 netmod=tcp pmi=pmi
```

What I get is
```
➜  ~ ls `spack location -i /4vs4lt3`/lib
libmpi.a    libmpicxx.a libpmpi.a   pkgconfig
➜  ~ ls `spack location -i /yw3oxjr`/lib
libmpi.12.dylib    libmpi.dylib       libmpichcxx.dylib  libmpicxx.a        libmpl.dylib       libpmpi.12.dylib   libpmpi.dylib
libmpi.a           libmpich.dylib     libmpicxx.12.dylib libmpicxx.dylib    libopa.dylib       libpmpi.a          pkgconfig
```

The conclusion of my tests is that shared libraries are not built if there is no fortran compiler.

Unfortunately I just have CMake knowledge and I know nothing abuot autotools, so I cannot try to fix it or investigate deeper at the moment. My speculation is that fortran checks "overwrites" previous checks about building shared libraries, and when there is no fortran compilers, the failure result in not building shared libraries. But again, it's just my speculation.

I'm sorry I'm not able to contribute more, but here it is the configure outputs for the two different builds.

[4vs4lt3-configure-out.txt](https://github.com/pmodels/mpich/files/7533901/spack-build-02-configure-out.txt)
[yw3oxjr-configure-out.txt](https://github.com/pmodels/mpich/files/7533903/spack-build-02-configure-out.txt)

**Comment by hzhou** [comment]
```
f77 = None
fc = None
```
This is suspicious. I don't the autotool handles `None` well. @albestro Do you the output from `make`?

**Comment by albestro** [comment]
Here it is the build output for the toolchain without the fortran compiler set. [4vs4lt3 spack-build-03-build-out.txt](https://github.com/pmodels/mpich/files/7534572/spack-build-03-build-out.txt)

I'm not sure what you mean with your comment; what you report is what spack sees, but then the build environment is arranged according to that configuration, but it does not mean that the fortran compilers are set to the string `None`.

This is how the build environment is set by spack when the fortran compiler is missing [spack-build-env.txt](https://github.com/pmodels/mpich/files/7534620/spack-build-env.txt).
Classic compilers variables are unset and their `SPACK_*` counterparts are populated. The latter ones are used by the spack compiler wrapper (https://github.com/spack/spack/blob/develop/lib/spack/env/cc).

Following your suggestion, I can see a possible breaking point in how the spack compiler wrapper manages the missing compiler

```
➜  ~ spack build-env mpich~fortran -- $SHELL
➜  ~ ftn
[spack cc] ERROR: Compiler 'apple-clang@13.0.0' does not have a Fortran 90 compiler configured.
```

Indeed, I don't know what autotools does exactly for checking the fortran compiler, but a missing executable can be interpreted differently from an executable that returns the above error string.

And if that's the case, I won't be so sure about what I stated in the beginning 😄 
> I decided to report here the issue because I think it is not related in any way to spack.

**Comment by hzhou** [comment]
In the `config.log` I see
```
114  checking for gfortran... gfortran
115  checking whether we are using the GNU Fortran compiler... no
...
```
It appears that spack is setting the fortran compiler to some non-functional dummy. Mpich should abort Fortran checking right there, but somehow didn't, then the logic goes into the weeds.

A correct spack formula should pass option `--disable-fortran` when Fortran compiler is not available, but the config logic in MPICH should also be improved.

**Comment by albestro** [comment]
> It appears that spack is setting the fortran compiler to some non-functional dummy.

Yes, exactly! It is what I meant in my previous comment: spack compiler wrapper acts as an existing compiler on the filesystem (because it can be executed), but as result it just prints out the error message
```
[spack cc] ERROR: Compiler 'apple-clang@13.0.0' does not have a Fortran 90 compiler configured.
```

> A correct spack formula should pass option `--disable-fortran` when Fortran compiler is not available, but the config logic in MPICH should also be improved.

It already does. Here it is the configuration parameters for that build.

`
--disable-silent-rules --enable-shared --with-hwloc-prefix=/Users/ialberto/spack/opt/spack/darwin-bigsur-skylake/apple-clang-13.0.0/hwloc-2.6.0-zruns75lvlkl6dqsjkcqop363tjfhm5g --with-pm=hydra --enable-romio --without-ibverbs --enable-wrapper-rpath=yes --disable-fortran --with-slurm=no --with-pmi=simple --with-device=ch3:nemesis:tcp --enable-libxml2 --enable-two-level-namespace
`

**Comment by hzhou** [comment]
I see. We need fix that.

**Comment by hzhou** [comment]
@albestro I just tried on my linux machine with hoaxing a `gfortran` using current `main` branch, and it builds and installed the shared library. I'll try the older release see if I can reproduce it.

**Comment by hzhou** [comment]
`3.4.2` also builds fine for me, need to look more into it.

**Comment by hzhou** [comment]
The suspicious line from config.log:
```
126  checking whether the gfortran linker (/Library/Developer/CommandLineTools/usr/bin/ld) supports shared libraries... no
```
Somehow the same linker `/Library/Developer/CommandLineTools/usr/bin/ld` returns `no` for this check, but in my testing it returned `yes`.

**Comment by albestro** [comment]
Thanks for the support @hzhou!

I think that it is a good idea to completely skip the check for fortran as you proposed in PR #5675.

Trying to give you useful information, I did another test by running the configure script in and out the spack build environment. In both cases a `gfortran` executable is available, but just in the latter one it is an actual fortran compiler.

Unfortunately, I was not able to identify the command that `configure` runs for this check, but running `configure` with the additional option `--enable-echo`, I was able to spot a potentially useful information: here it is reported the diff of the same excerpt from the two different outputs (left: inside spack build env / right: outside spack build env).

```
+ case $cc_basename in						+ case $cc_basename in
+ _lt_dar_can_shared=					      |	+ _lt_dar_can_shared=yes
+ test yes = ''						      |	+ test yes = yes
+ ld_shlibs_F77=no					      |	+ output_verbose_link_cmd=func_echo_all
							      >	+ archive_cmds_F77='$CC -dynamiclib $allow_undefined_flag -o
							      >	+ module_cmds_F77='$CC $allow_undefined_flag -o $lib -bundle
							      >	+ archive_expsym_cmds_F77='sed '\''s|^|_|'\'' < $export_symbo
							      >	+ module_expsym_cmds_F77='sed -e '\''s|^|_|'\'' < $export_sym
+ test sni = apple						+ test sni = apple
+ printf '%s\n' 'configure:25238: result: no'		      |	+ printf '%s\n' 'configure:25238: result: yes'
+ printf '%s\n' no					      |	+ printf '%s\n' yes
no							      |	yes
```

I tried to investigate a bit inside the configure script (unfortunately I cannot link it via github because in the repo there are just generator scripts, I'm referring to the configure script available in the tarball for 3.4.2), to see where those lines come from, and I think they refer to this snippet.

```
24794   case $cc_basename in
24795      ifort*|nagfor*)
24796         lt_prog_compiler_wl_F77='-Wl,'
24797         _lt_dar_can_shared=yes
24798       ;;
24799      *) _lt_dar_can_shared=$GCC ;;
24800   esac
24801   if test yes = "$_lt_dar_can_shared"; then
24802     output_verbose_link_cmd=func_echo_all
24803     archive_cmds_F77="\$CC -dynamiclib \$allow_undefined_flag -o \$lib \$libobjs \$deplibs \$compiler_flags -install_name \$rpath/\$soname \$verstring $_lt_dar_
24804     module_cmds_F77="\$CC \$allow_undefined_flag -o \$lib -bundle \$libobjs \$deplibs \$compiler_flags$_lt_dsymutil"
24805     archive_expsym_cmds_F77="sed 's|^|_|' < \$export_symbols > \$output_objdir/\$libname-symbols.expsym~\$CC -dynamiclib \$allow_undefined_flag -o \$lib \$libob
24806     module_expsym_cmds_F77="sed -e 's|^|_|' < \$export_symbols > \$output_objdir/\$libname-symbols.expsym~\$CC \$allow_undefined_flag -o \$lib -bundle \$libobjs
24807
24808   else
24809   ld_shlibs_F77=no
24810   fi
```

In particular, I would say that using spack build environment, line 24799 is executed giving  `_lt_dar_can_shared=''` and so resulting in `ld_shlibs_F77=no`, which triggers the problem.

I want to highlight that I did another test: by removing `gfortran` from my PATH, this check do not appear at all: probably, by not finding the executable, these tests are completely skipped.
